File: Data Structures & Algorithms/sliding-window-maximum/submission-2.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        TreeMap<Integer, Integer> window = new TreeMap<>();
        int[] result = new int[nums.length-k+1];
        for (int i=0; i<k; i++) {
            window.put(nums[i], window.getOrDefault(nums[i], 0) + 1);
        }
        result[0] = window.lastKey();
        for (int i=k; i<nums.length; i++) {
            window.put(nums[i], window.getOrDefault(nums[i], 0) + 1);
            window.put(nums[i-k], window.getOrDefault(nums[i-k], 0) - 1);
            if (window.get(nums[i-k]) == 0) window.remove(nums[i-k]);
            result[i-k+1] = window.lastKey();
        }
        return result;
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        foundZero = False
        product = 1
        for num in nums:
            if num == 0:
                if foundZero == True:
                    return [0] * len(nums)
                foundZero = True
            else:
                product *= num
        result = []
        for num in nums:
            if num == 0:
                result.append(product)
            elif foundZero:
                result.append(0)
            else:
                result.append(product // num)
        return result
